containers/src/Kokkos_DynamicView.hpp

```diff
@@ -53,36 +53,184 @@
 namespace Kokkos {
 namespace Experimental {
 
-// Simple metafunction for choosing memory space
-// In the current implementation, if memory_space == CudaSpace,
-// use CudaUVMSpace for the chunk 'array' allocation, which
-// contains will contain pointers to chunks of memory allocated
-// in CudaSpace
 namespace Impl {
-template <class MemSpace>
-struct ChunkArraySpace {
-  using memory_space = MemSpace;
-};
 
-#ifdef KOKKOS_ENABLE_CUDA
-template <>
-struct ChunkArraySpace<Kokkos::CudaSpace> {
-  using memory_space = typename Kokkos::CudaUVMSpace;
-};
-#endif
-#ifdef KOKKOS_ENABLE_HIP
-template <>
-struct ChunkArraySpace<Kokkos::Experimental::HIPSpace> {
-  using memory_space = typename Kokkos::Experimental::HIPHostPinnedSpace;
+template <typename MemorySpace, typename ValueType>
+struct Destroy {
+  using value_type = ValueType;
+
+  Destroy()               = default;
+  Destroy(Destroy&&)      = default;
+  Destroy(const Destroy&) = default;
+  Destroy& operator=(Destroy&&) = default;
+  Destroy& operator=(const Destroy&) = default;
+
+  Destroy(std::string label, ValueType** arg_chunk,
+          const unsigned arg_chunk_max, const unsigned arg_chunk_size,
+          value_type** arg_linked)
+      : m_label(label),
+        m_chunks(arg_chunk),
+        m_linked(arg_linked),
+        m_chunk_max(arg_chunk_max),
+        m_chunk_size(arg_chunk_size) {}
+
+  void execute() {
+    // Destroy the array of chunk pointers.
+    // Two entries beyond the max chunks are allocation counters.
+    for (unsigned i = 0; i < m_chunk_max; i++) {
+      MemorySpace().deallocate(m_label.c_str(), m_chunks[i],
+                               sizeof(value_type) * m_chunk_size);
+    }
+    // Destroy the linked allocate if we have one.
+    if (m_linked != nullptr) {
+      MemorySpace().deallocate(m_label.c_str(), m_linked,
+                               (sizeof(value_type*) * (m_chunk_max + 2)));
+    }
+  }
+
+  void destroy_shared_allocation() { execute(); }
+
+  std::string m_label;
+  value_type** m_chunks = nullptr;
+  value_type** m_linked = nullptr;
+  unsigned m_chunk_max;
+  unsigned m_chunk_size;
 };
-#endif
-#ifdef KOKKOS_ENABLE_SYCL
-template <>
-struct ChunkArraySpace<Kokkos::Experimental::SYCLDeviceUSMSpace> {
-  using memory_space = typename Kokkos::Experimental::SYCLSharedUSMSpace;
+
+template <typename MemorySpace, typename ValueType>
+struct SmartMemoryAccessor {
```

Comment:
short documentation comment would be good. This is mostly a utility class for arrays of pointers or?

Reply:
Yes, I've added a comment to explain what it does. I can't move it out of the `Impl` namespace because it's a subobject of `DynamicView` so it can't be in an anon namespace.